pkg/pillar/cmd/zedrouter/appnumonunet.go

```diff
@@ -0,0 +1,262 @@
+		appNum = 0
+		log.Functionf("Allocating appNum %d for %s isZedmanager",
+			appNum, appNumOnUNetKey(baseID, appID))
+	} else {
+		// XXX could look for non-0xFF bytes first for efficiency
+		appNum = -1
+		if isStatic {
+			for i := types.BitMapMax; i >= 0; i-- {
+				if !baseMap.IsSet(i) {
+					appNum = i
+					log.Functionf("Allocating appNum %d for %s",
+						appNum, appNumOnUNetKey(baseID, appID))
+					break
+				}
+			}
+		} else {
+			for i := 0; i <= types.BitMapMax; i++ {
+				if !baseMap.IsSet(i) {
+					appNum = i
+					log.Functionf("Allocating appNum %d for %s",
+						appNum, appNumOnUNetKey(baseID, appID))
+					break
+				}
+			}
+		}
+		if appNum == -1 {
+			log.Functionf("Failed to find free appNum for %s. Reusing!",
+				appNumOnUNetKey(baseID, appID))
+			oldAppID, oldAppNum, err :=
+				uuidpairtonum.NumGetOldestUnused(log, pub,
+					baseID, numType)
+			if err != nil {
+				errStr := fmt.Sprintf("no free appNum")
+				log.Fatal(errStr)
+			}
+			log.Functionf("Reuse found appNum %d for %s. Reusing!",
+				oldAppNum, appNumOnUNetKey(baseID, oldAppID))
+			uuidpairtonum.NumDelete(log, pub, baseID, oldAppID)
+			baseMap.Clear(oldAppNum)
+			appNum = oldAppNum
+		}
+	}
+	if baseMap.IsSet(appNum) {
+		log.Fatalf("Bitmap is already set for %d", appNum)
+	}
+	baseMap.Set(appNum)
+	uuidpairtonum.NumAllocate(log, pub, baseID, appID, appNum, true,
+		numType)
+	return appNum, nil
+}
+
+func appNumOnUNetFree(ctx *zedrouterContext, baseID uuid.UUID,
+	appID uuid.UUID) {
+
+	pub := ctx.pubUUIDPairToNum
+	numType := appNumOnUNetType
+	appNum, err := uuidpairtonum.NumGet(log, pub, baseID, appID, numType)
+	if err != nil {
+		log.Fatalf("num not found for %s",
+			appNumOnUNetKey(baseID, appID))
+	}
+	baseMap := appNumOnUNetBaseGet(baseID)
+	if baseMap == nil {
+		uuidpairtonum.NumDelete(log, pub, baseID, appID)
+		return
+	}
+	// Check that number exists in the allocated numbers
+	if !baseMap.IsSet(appNum) {
+		log.Fatalf("Bitmap is not set for %d", appNum)
+	}
+	baseMap.Clear(appNum)
+	uuidpairtonum.NumDelete(log, pub, baseID, appID)
+}
+
+func appNumOnUNetGet(ctx *zedrouterContext, baseID uuid.UUID,
+	appID uuid.UUID) (int, error) {
+	pub := ctx.pubUUIDPairToNum
+	numType := appNumOnUNetType
+	return uuidpairtonum.NumGet(log, pub, baseID, appID, numType)
+}
+
+// returns base bitMap for a given UUID
+func appNumOnUNetBaseGet(baseID uuid.UUID) *types.Bitmap {
+	if baseMap, exist := appNumBase[baseID.String()]; exist {
+		return baseMap
+	}
+	return nil
+}
+
+// Create application number Base for a given UUID
+func appNumOnUNetBaseCreate(baseID uuid.UUID) *types.Bitmap {
+	if appNumOnUNetBaseGet(baseID) == nil {
+		log.Functionf("appNumOnUNetBaseCreate (%s)", baseID.String())
+		appNumBase[baseID.String()] = new(types.Bitmap)
+	}
+	return appNumOnUNetBaseGet(baseID)
+}
+
+// Delete the application number Base for a given UUID
+func appNumOnUNetBaseDelete(baseID uuid.UUID) {
+	if appNumOnUNetBaseGet(baseID) != nil {
```

Comment:
See above - this should fatal if nil because that indicates that something is seriously broken.